File: examples/2015-day-01/ocaml/lib/aoc.ml
let step floor instruction =
  match instruction with
  | '(' -> floor + 1
  | ')' -> floor - 1
  | _ -> floor
;;

let solve_part_one input_text = input_text |> String.to_seq |> Seq.fold_left step 0

let solve_part_two input_text =
  input_text
  |> String.to_seq
  |> Seq.scan step 0
  |> Seq.drop 1
  |> Seq.find_index (fun floor -> floor = -1)
  |> Option.map (fun index -> index + 1)
  |> Option.value ~default:0
;;

File: examples/2015-day-01/fsharp/src/Aoc.fs
module Aoc

let step floor instruction =
    match instruction with
    | '(' -> floor + 1
    | ')' -> floor - 1
    | _ -> invalidArg (nameof instruction) $"Invalid instruction '{instruction}', expected '(' or ')'"

/// Solves part one of the puzzle for the given input.
let solvePartOne (input : string) : int = input |> Seq.fold step 0

/// Solves part two of the puzzle for the given input.
let solvePartTwo (input : string) : int =
    input
    |> Seq.scan step 0
    |> Seq.skip 1
    |> Seq.tryFindIndex (fun floor -> floor = -1)
    |> Option.map (fun index -> index + 1)
    |> Option.defaultValue 0
